src/Symfony/Component/HttpKernel/HttpCache/Psr6Store.php

```diff
@@ -0,0 +1,616 @@
+     *
+     * @return string The key under which the response is stored
+     */
+    public function write(Request $request, Response $response): string
+    {
+        if (null === $response->getMaxAge()) {
+            throw new \InvalidArgumentException('HttpCache should not forward any response without any cache expiration time to the store.');
+        }
+
+        // Save the content digest if required
+        $this->saveContentDigest($response);
+
+        $cacheKey = $this->getCacheKey($request);
+        $headers = $response->headers->all();
+        unset($headers['age']);
+
+        $item = $this->cache->getItem($cacheKey);
+
+        if (!$item->isHit()) {
+            $entries = [];
+        } else {
+            $entries = $item->get();
+        }
+
+        // Add or replace entry with current Vary header key
+        $varyKey = $this->getVaryKey($response->getVary(), $request);
+        $entries[$varyKey] = [
+            'vary' => $response->getVary(),
+            'headers' => $headers,
+            'status' => $response->getStatusCode(),
+            'uri' => $request->getUri(), // For debugging purposes
+        ];
+
+        // Add content if content digests are disabled
+        if (!$this->options['generate_content_digests']) {
+            $entries[$varyKey]['content'] = $response->getContent();
+        }
+
+        // If the response has a Vary header we remove the non-varying entry
+        if ($response->hasVary()) {
+            unset($entries[self::NON_VARYING_KEY]);
+        }
+
+        // Tags
+        $tags = [];
+        foreach ($response->headers->all($this->options['cache_tags_header']) as $header) {
+            foreach (explode(',', $header) as $tag) {
+                $tags[] = $tag;
+            }
+        }
+
+        // Prune expired entries on file system if needed
+        $this->autoPruneExpiredEntries();
+
+        $this->saveDeferred($item, $entries, $response->getMaxAge(), $tags);
+
+        // Commit all deferred cache items
+        $this->cache->commit();
+
+        return $cacheKey;
+    }
+
+    /**
+     * Invalidates all cache entries that match the request.
+     *
+     * @param Request $request A Request instance
+     */
+    public function invalidate(Request $request): void
+    {
+        $cacheKey = $this->getCacheKey($request);
+
+        $this->cache->deleteItem($cacheKey);
+    }
+
+    /**
+     * Locks the cache for a given Request.
+     *
+     * @param Request $request A Request instance
+     *
+     * @return bool|string true if the lock is acquired, the path to the current lock otherwise
+     */
+    public function lock(Request $request)
+    {
+        $cacheKey = $this->getCacheKey($request);
+
+        if (isset($this->locks[$cacheKey])) {
+            return false;
+        }
+
+        $this->locks[$cacheKey] = $this->lockFactory
+            ->createLock($cacheKey);
+
+        return $this->locks[$cacheKey]->acquire();
```

Comment:
if lock is not acquired, should you unset `this->locks[$cacheKey]`?

one could retry to lock the resource several time until it works